### ISSUE TITLE
v4 alpha3 - Problem backing up a directory with large number of files with rsync

### ISSUE THREAD

**Issue by adamgoryachev** [backuppc/backuppc#1]
Copied from an email sent to the BPC user list on 2015-07-20:
 BackupPC then tries to backup the host using rsync over SSH

The problem seems to be that there is a directory with over 700,000 files (images) in it.

This seems to be a large part of the problem (taken from /var/log/kern.log)
Jul 19 08:53:13 keep kernel: [22558938.642495] rsync_bpc[1653]: segfault at 7fc12f007928 ip 0000000000449333 sp 00007fff79684680 error 4 in rsync_bpc[400000+70000]
Jul 19 10:26:55 keep kernel: [22564550.290226] rsync_bpc[8905]: segfault at 7f3a476952a8 ip 0000000000449333 sp 00007fff23fb6e10 error 4 in rsync_bpc[400000+70000]
Jul 19 12:26:47 keep kernel: [22571728.343674] rsync_bpc[17464]: segfault at 7fd4184fbd28 ip 0000000000449333 sp 00007ffff5fbe7c0 error 4 in rsync_bpc[400000+70000]
Jul 19 13:57:40 keep kernel: [22577171.017578] rsync_bpc[23984]: segfault at 7fee9daa59a8 ip 0000000000449333 sp 00007fffa00445b0 error 4 in rsync_bpc[400000+70000]
Jul 19 14:43:35 keep kernel: [22579919.766906] rsync_bpc[27380]: segfault at 7f17ff2df5a8 ip 0000000000449333 sp 00007fff2a16b750 error 4 in rsync_bpc[400000+70000]
Jul 19 15:41:02 keep kernel: [22583360.800624] rsync_bpc[31446]: segfault at 7f5e14f6e5a8 ip 0000000000449333 sp 00007fff8096fd80 error 4 in rsync_bpc[400000+70000]
Jul 19 17:10:34 keep kernel: [22588721.803461] rsync_bpc[2352]: segfault at 7f7cc61605a8 ip 0000000000449333 sp 00007fff567451d0 error 4 in rsync_bpc[400000+70000]
Jul 19 17:44:27 keep kernel: [22590751.533047] rsync_bpc[8697]: segfault at 7f53b7f34928 ip 0000000000449333 sp 00007fff6f146d20 error 4 in rsync_bpc[400000+70000]
Jul 19 18:12:13 keep kernel: [22592413.839811] rsync_bpc[11032]: segfault at 7f9dc452fd28 ip 0000000000449333 sp 00007fffae9d27f0 error 4 in rsync_bpc[400000+70000]
Jul 19 18:40:14 keep kernel: [22594091.148687] rsync_bpc[13181]: segfault at 7f68d9d1cb28 ip 0000000000449333 sp 00007fffb6897070 error 4 in rsync_bpc[400000+70000]
Jul 19 19:07:45 keep kernel: [22595738.996516] rsync_bpc[15129]: segfault at 7f2cdc961a28 ip 0000000000449333 sp 00007fff3fbb3100 error 4 in rsync_bpc[400000+70000]
Jul 19 19:34:52 keep kernel: [22597363.711624] rsync_bpc[17262]: segfault at 7f7974af4128 ip 0000000000449333 sp 00007fffbab514a0 error 4 in rsync_bpc[400000+70000]
Jul 20 01:39:48 keep kernel: [22619216.885463] rsync_bpc[10515]: segfault at 7fc5e009dca8 ip 0000000000449333 sp 00007fff4845e9c0 error 4 in rsync_bpc[400000+70000]
Jul 20 10:09:32 keep kernel: [22649741.538150] rsync_bpc[13981]: segfault at 7fc96a14b728 ip 0000000000449333 sp 00007fff5bda86e0 error 4 in rsync_bpc[400000+70000]

**Comment by adamgoryachev** [comment]
Found my post from 2016-01-18 with more detailed analysis:
Jan 18 13:15:05 keep kernel: [11562928.034694] rsync_bpc[1224]: segfault 
at 7f6923390428 ip 00000000004473af sp 00007fff7d94c8f0 error 4 in 
rsync_bpc[400000+75000]
Jan 18 13:30:57 keep kernel: [11563881.316870] rsync_bpc[1322]: segfault 
at 7f8a9f83b428 ip 00000000004473af sp 00007fff9b9d9850 error 4 in 
rsync_bpc[400000+75000]

I've found an informative post: 
http://stackoverflow.com/questions/2549214/interpreting-segfault-messages and 
this pointed me to this command:
addr2line -e /usr/local/bin/rsync_bpc -fCi 0x00000000004473af
bpc_attrib_fileCopyOpt
/usr/src/rsync-bpc-3.0.9.3/backuppc/bpc_attrib.c:284

Looking at the file I see this function:
    273  /*
    274   * Copy all the attributes from fileSrc to fileDest. fileDest 
should already have a
    275   * valid allocated fileName and allocated xattr hash.  The 
fileDest xattr hash is
    276   * emptied before the copy, meaning it is over written.
    277   *
    278   * If overwriteEmptyDigest == 0, an empty digest in fileSrc 
will not overwrite fileDest.
    279   */
    280  void bpc_attrib_fileCopyOpt(bpc_attrib_file *fileDest, 
bpc_attrib_file *fileSrc, int overwriteEmptyDigest)
    281  {
    282      if ( fileDest == fileSrc ) return;
    283
    284      fileDest->type      = fileSrc->type;
    285      fileDest->compress  = fileSrc->compress;
    286      fileDest->mode      = fileSrc->mode;
    287      fileDest->isTemp    = fileSrc->isTemp;
    288      fileDest->uid       = fileSrc->uid;
    289      fileDest->gid       = fileSrc->gid;
    290      fileDest->nlinks    = fileSrc->nlinks;
    291      fileDest->mtime     = fileSrc->mtime;
    292      fileDest->size      = fileSrc->size;
    293      fileDest->inode     = fileSrc->inode;

Looking at line 284 we see that this is the first time we try to read 
from the object fileSrc. I suspect that somehow fileSrc is either 
invalid, doesn't exist, etc, and therefore that is why we are getting 
this error. I'm guessing that is has some value, as otherwise we should 
see a much smaller number in the at value from the logs (similar to the 
OP in the stackoverflow message.

So, is there a simple way to make sure fileSrc is "valid" before trying 
to read it, and potentially cause a crash? I'd like to add some extra 
logs/debug before the crash to try and find the cause and hopefully fix it.

There are only two places that this function is called:
/*
  * Copy all the attributes from fileSrc to fileDest.  fileDest should 
already have a
  * valid allocated fileName and allocated xattr hash.  The fileDest 
xattr hash is
  * emptied before the copy, meaning it is over written.
  */
void bpc_attrib_fileCopy(bpc_attrib_file *fileDest, bpc_attrib_file 
*fileSrc)
{
     if ( fileDest == fileSrc ) return;

     bpc_attrib_fileCopyOpt(fileDest, fileSrc, 1);
}

This is just passing the exact same variable it received, so we will 
need to trace it back another step... I guess if there is an easy method 
to test if it is valid, I can add that test to each function before 
calling bpc_attrib_fileCopy and hopefully eventually work out what is 
wrong with it.

I guess my c skills are quite rusty (non-existant really), so if anyone 
is able to assist, I'd be very happy, even if it is just a clue on the 
right way to debug/find the problem.
